### ISSUE TITLE
RNN op with dropout cannot use fixed seed on CPU

### ISSUE THREAD

**Issue by roywei** [apache/incubator-mxnet#16604]
As follow up on https://github.com/apache/incubator-mxnet/pull/16532

I found out rnn operator with dropout result cannot be fixed with fixed seed on CPU.
GPU (cudnn) is fine after PR https://github.com/apache/incubator-mxnet/pull/16532

```
@with_seed()
def test_rnn_with_seed():
    info = np.iinfo(np.int32)
    seed = np.random.randint(info.min, info.max)
    
    _test_rnn(seed, mx.cpu())
    _test_rnn(seed, mx.gpu())

def _test_rnn(seed, ctx):
    data = mx.nd.ones((5, 3, 10), ctx=ctx)
    rnn = mx.gluon.rnn.RNN(100, 3, dropout=0.5)
    rnn.initialize(ctx=ctx)
    mx.random.seed(seed)
    with mx.autograd.record():
        result1 = rnn(data)

    mx.random.seed(seed)
    with mx.autograd.record():
        result2 = rnn(data)
    # dropout on gpu should return same result with fixed seed
    assert_almost_equal(result1.asnumpy(), result2.asnumpy())

```

Current workaround is DO NOT use dropout in rnn and manually append a dropout layer after rnn, the following works:
```
def _test_rnn(seed, ctx):
    data = mx.nd.ones((5, 3, 10), ctx=ctx)
    rnn = mx.gluon.rnn.RNN(100, 3, dropout=0.)
    rnn.initialize(ctx=ctx)
    dropout = mx.gluon.nn.Dropout(0.5)
    with mx.autograd.record():
        result1 = rnn(data)
        mx.random.seed(seed)
        o1 = dropout(result1)

    with mx.autograd.record():
        result2 = rnn(data)
        mx.random.seed(seed)
        o2 = dropout(result2)
    # dropout on gpu should return same result with fixed seed
    assert_almost_equal(o1.asnumpy(), o2.asnumpy())
```

**Comment by roywei** [comment]
This implementation on lstm, gru and rnn should respect mxnet seed:

https://github.com/apache/incubator-mxnet/blob/master/src/operator/rnn_impl.h#L159

**Comment by pengzhao-intel** [comment]
@zixuanweeei please take a look for this

**Comment by zixuanweeei** [comment]
> @zixuanweeei please take a look the fix from @roywei :)

Sure. And thanks for @roywei 's great work.

**Comment by samskalicky** [comment]
@lanking520 assign @zixuanweeei 
Any update on this?

**Comment by roywei** [comment]
suggest to use workaround provided in issue for now, proper fix depends on fix on dropout operator randomness on both CPU and GPU, which will take longer time.

**Comment by zixuanweeei** [comment]
Is there any issue or proposal tracking the problem among the randomness of the dropout operator. It would be highly appreciated if you could provide some information for us. And what is the challenge in fixing the randomness? Thanks. @roywei

**Comment by roywei** [comment]
@zixuanweeei  the problems are discussed in this PR: https://github.com/apache/incubator-mxnet/pull/16532 and this issue: https://github.com/apache/incubator-mxnet/issues/16705

the problem is mainly on GPU when we call `CUDNN_CALL(cudnnSetDropoutDescriptor)`, every time we call ` mx.random.seed(x)`, we need to figure out a way to clear dropout states on GPU and call `cudnnSetDropoutDescriptor` again to set a new seed from MXNet random module.  right now dropout op only call `cudnnSetDropoutDescriptor`  once at the beginning, rest of the time it goes to `cudnnRestoreDropoutDescriptor`

**Comment by roywei** [comment]
On second though, maybe I can fix rnn on CPU first, I will leave GPU logic the same

**Comment by zixuanweeei** [comment]
> On second though, maybe I can fix rnn on CPU first, I will leave GPU logic to the original implementation here: https://github.com/apache/incubator-mxnet/pull/16605/files#diff-6dfdca409e69cc495f286170fe1e553eL1362

Thanks for the elaboration. Besides, there are also some problems sharing the same seed among multi-GPU, right? I agree with you that we can fix rnn on CPU first :slightly_smiling_face:.
